fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java

```diff
@@ -431,10 +440,85 @@ protected void handleResultsResponse(
             );
             return Observable.error(responseException);
         }
+
+        if (jobsApiRequest.getPaginationParameters().isPresent()) {
+            Pagination<Result> pages = new AllPagesPagination<>(
+                    preResponse.getResultSet(),
+                    jobsApiRequest.getPaginationParameters().get()
+            );
+            responseContext = paginateResponseContext(responseContext, jobsApiRequest, pages);
+            PreResponse paginatedPreResponse = new PreResponse(
+                    new ResultSet(pages.getPageOfData(), preResponse.getResultSet().getSchema()),
+                    responseContext
+            );
+            return Observable.just(paginatedPreResponse);
+        }
         return Observable.just(preResponse);
     }
 
     /**
+     * Add pagination details to ResponseContext.
+     *
+     * @param responseContext  ResponseContext object contains all the meta info of the resultSet
+     * @param jobsApiRequest  JobsApiRequest object with all the associated info with it
+     * @param pages  Paginated resultSet
+     *
+     * @return Updated ResponseContext contains pagination info
+     */
+    protected ResponseContext paginateResponseContext(
+            ResponseContext responseContext,
+            ApiRequest jobsApiRequest,
+            Pagination<Result> pages
+    ) {
+        LinkedHashMap<String, URI> bodyLinks = new LinkedHashMap<>();
+        Arrays.stream(PaginationLink.values())
+                .forEachOrdered(link -> addLink(
+                                link,
+                                pages,
+                                jobsApiRequest.getUriInfo().getRequestUriBuilder(),
+                                bodyLinks)
+                );
```

Comment:
I'm pretty sure you copy-pasted this from the `ApiRequest`, which does the same thing, which is fine.

Except that this code is ugly, as far as I'm concerned, and doesn't really use streams. I also find it hard to read because we have an overloaded method `addLink` that do rather different things. Instead, I would propose that we take more full advantage of streams, and do the following:

```java
LinkedHashMap<String, URI> bodyLinks = Arrays.stream(PaginationLink.values())
                .map(link -> new ImmutablePair<>(link.getBodyName(),link.getPage(pages)))
                .filter(pair -> pair.getRight().isPresent())
                .map(pair -> Utils.withRight(pair, pair.getRight().getAsInt()))
                .map(pair -> Utils.withRight(pair, uriBuilder.replaceQueryParam("page", pair.getRight())))
                .map(pair -> Utils.withRight(pair, pair.getRight().build()))
                .collect(StreamUtils.toLinkedMap(Pair::getLeft, Pair::getRight));
```

A few comments: The `ImmutablePair` class is from the apache commons. I also added a utility method `withRight` with the expected semantics:

```java
public static <T, U, V> ImmutablePair<T, V> withRight(ImmutablePair<T, U> pair, V right) {
        return new ImmutablePair<>(pair.getLeft(), right);
}
```
And a similar one for replacing the left value.